File: src/main/java/ui_adapter/testcase/ExcelTestCaseLoader.java
package ui_adapter.testcase;

import org.apache.poi.ss.usermodel.*;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;
import ui_adapter.model.TestCase;
import ui_adapter.model.TestStep;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.*;

/**
 * Loads test cases from Excel files.
 * Supports reading test cases with multiple steps from Excel format.
 */
public class ExcelTestCaseLoader implements TestCaseLoader {

    // Column names as they appear in the Excel file
    private static final String COL_ISSUE_KEY = "Issue Key";
    private static final String COL_SUMMARY = "Summary";
    private static final String COL_DESCRIPTION = "Description";
    private static final String COL_PRECONDITION = "Precondition";
    private static final String COL_STATUS = "Status";
    private static final String COL_PRIORITY = "Priority";
    private static final String COL_STEP_SUMMARY = "Step Summary";
    private static final String COL_TEST_STEP = "Test Step";  // Alternative name for Step Summary
    private static final String COL_TEST_DATA = "Test Data";
    private static final String COL_EXPECTED_RESULT = "Expected Result";
    private static final String COL_TESTCASE_TYPE = "TestCase Type";
    private static final String COL_FUNCTIONAL_CATEGORY = "functional category";
    private static final String COL_COMPONENT_UNDER_TEST = "component under test";

    private final String filePath;

    public ExcelTestCaseLoader(String filePath) {
        this.filePath = filePath;
    }

    @Override
    public List<TestCase> loadTestCases() throws IOException {
        List<TestCase> testCases = new ArrayList<>();
        
        File file = new File(filePath);
        if (!file.exists()) {
            throw new IOException("Excel file not found: " + filePath);
        }

        try (FileInputStream fis = new FileInputStream(file);
             Workbook workbook = new XSSFWorkbook(fis)) {
            
            Sheet sheet = workbook.getSheetAt(0); // Read first sheet
            
            // Read header row to get column indices
            Row headerRow = sheet.getRow(0);
            if (headerRow == null) {
                throw new IOException("Excel file is empty or has no header row");
            }
            
            Map<String, Integer> columnMap = createColumnMap(headerRow);
            
            // Read data rows
            Map<String, TestCase> testCaseMap = new LinkedHashMap<>();
            String lastIssueKey = null;
            
            for (int i = 1; i <= sheet.getLastRowNum(); i++) {
                Row row = sheet.getRow(i);
                if (row == null || isEmptyRow(row)) {
                    continue;
                }
                
                String issueKey = getCellValue(row, columnMap.get(COL_ISSUE_KEY));
                
                // If issue key is empty, try to use the last seen issue key (continuation of test case)
                if (issueKey == null || issueKey.trim().isEmpty()) {
                    if (lastIssueKey != null) {
                        issueKey = lastIssueKey;
                    } else {
                        // Skip if no issue key and no previous issue key
                        continue;
                    }
                } else {
                    lastIssueKey = issueKey;
                }
                
                // Get or create test case
                TestCase testCase = testCaseMap.get(issueKey);
                if (testCase == null) {
                    testCase = new TestCase();
                    testCase.setKey(issueKey);
                    testCase.setSummary(getCellValue(row, columnMap.get(COL_SUMMARY)));
                    testCase.setDescription(getCellValue(row, columnMap.get(COL_DESCRIPTION)));
                    testCase.setPreconditions(getCellValue(row, columnMap.get(COL_PRECONDITION)));
                    testCase.setStatus(getCellValue(row, columnMap.get(COL_STATUS)));
                    testCase.setPriority(getCellValue(row, columnMap.get(COL_PRIORITY)));
                    testCase.setTestCaseType(getCellValue(row, columnMap.get(COL_TESTCASE_TYPE)));
                    testCase.setFunctionalCategory(getCellValue(row, columnMap.get(COL_FUNCTIONAL_CATEGORY)));
                    testCase.setComponentUnderTest(getCellValue(row, columnMap.get(COL_COMPONENT_UNDER_TEST)));
                    
                    testCaseMap.put(issueKey, testCase);
                }
                
                // Add test step if any step data exists
                // Support both "Step Summary" and "Test Step" column names
                String stepSummary = getCellValue(row, columnMap.get(COL_STEP_SUMMARY));
                if (stepSummary == null || stepSummary.trim().isEmpty()) {
                    stepSummary = getCellValue(row, columnMap.get(COL_TEST_STEP));
                }
                
                String testData = getCellValue(row, columnMap.get(COL_TEST_DATA));
                String expectedResult = getCellValue(row, columnMap.get(COL_EXPECTED_RESULT));
                
                // Add step if there's at least an action or expected result
                if ((stepSummary != null && !stepSummary.trim().isEmpty()) ||
                    (expectedResult != null && !expectedResult.trim().isEmpty())) {
                    
                    TestStep testStep = new TestStep(stepSummary, testData, expectedResult);
                    testCase.addTestStep(testStep);
                }
            }
            
            testCases.addAll(testCaseMap.values());
        }
        
        return testCases;
    }

    @Override
    public TestCase loadTestCase(String testCaseKey) throws IOException {
        List<TestCase> allTestCases = loadTestCases();
        
        for (TestCase testCase : allTestCases) {
            if (testCase.getKey() != null && testCase.getKey().equals(testCaseKey)) {
                return testCase;
            }
        }
        
        throw new IOException("Test case not found with key: " + testCaseKey);
    }

    /**
     * Creates a map of column names to their indices
     */
    private Map<String, Integer> createColumnMap(Row headerRow) {
        Map<String, Integer> columnMap = new HashMap<>();
        
        for (int i = 0; i < headerRow.getLastCellNum(); i++) {
            Cell cell = headerRow.getCell(i);
            if (cell != null) {
                String columnName = cell.getStringCellValue().trim();
                columnMap.put(columnName, i);
            }
        }
        
        return columnMap;
    }

    /**
     * Gets cell value as string, handling different cell types
     */
    private String getCellValue(Row row, Integer columnIndex) {
        if (columnIndex == null) {
            return null;
        }
        
        Cell cell = row.getCell(columnIndex);
        if (cell == null) {
            return null;
        }
        
        switch (cell.getCellType()) {
            case STRING:
                return cell.getStringCellValue();
            case NUMERIC:
                if (DateUtil.isCellDateFormatted(cell)) {
                    return cell.getDateCellValue().toString();
                } else {
                    return String.valueOf(cell.getNumericCellValue());
                }
            case BOOLEAN:
                return String.valueOf(cell.getBooleanCellValue());
            case FORMULA:
                try {
                    return cell.getStringCellValue();
                } catch (IllegalStateException e) {
                    return String.valueOf(cell.getNumericCellValue());
                }
            case BLANK:
                return null;
            default:
                return null;
        }
    }

    /**
     * Checks if a row is empty
     */
    private boolean isEmptyRow(Row row) {
        for (int i = 0; i < row.getLastCellNum(); i++) {
            Cell cell = row.getCell(i);
            if (cell != null && cell.getCellType() != CellType.BLANK) {
                String value = getCellValue(row, i);
                if (value != null && !value.trim().isEmpty()) {
                    return false;
                }
            }
        }
        return true;
    }

    /**
     * Utility method to print test case summary for debugging
     */
    public static void printTestCaseSummary(TestCase testCase) {
        System.out.println("=".repeat(80));
        System.out.println("Test Case: " + testCase.getKey());
        System.out.println("Summary: " + testCase.getSummary());
        System.out.println("Preconditions: " + testCase.getPreconditions());
        System.out.println("Number of Steps: " + testCase.getTestSteps().size());
        System.out.println("=".repeat(80));
        System.out.println();
    }
}

File: python-code-base/playwright_ui_adapter/testcase/excel_test_case_loader.py
"""Excel test case loader implementation."""

from pathlib import Path
from typing import List, Dict, Any, Optional

import openpyxl
from openpyxl.worksheet.worksheet import Worksheet

from ..model.test_case import TestCase
from ..model.test_step import TestStep
from .test_case_loader import TestCaseLoader


class ExcelTestCaseLoader(TestCaseLoader):
    """
    Loads test cases from Excel files.
    
    Expected Excel format:
    - Each row represents a test step
    - Rows with the same "Issue Key" are grouped into one test case
    - Required columns: Issue Key, Test Summary, Test Step, Expected Result
    - Optional columns: Preconditions, Test Data, Priority, etc.
    
    Example:
        >>> loader = ExcelTestCaseLoader()
        >>> test_cases = loader.load_test_cases(Path("testcases.xlsx"))
        >>> for test_case in test_cases:
        ...     print(f"Test: {test_case.summary}")
    """
    
    # Default column mappings (can be customized)
    DEFAULT_COLUMNS = {
        "issue_key": "Issue Key",
        "summary": "Test Summary",
        "preconditions": "Preconditions",
        "test_step": "Test Step",
        "test_data": "Test Data",
        "expected_result": "Expected Result",
        "description": "Description",
        "status": "Status",
        "priority": "Priority",
        "test_case_type": "Test Case Type",
        "functional_category": "Functional Category",
        "component": "Component Under Test",
    }
    
    def __init__(self, column_mapping: Optional[Dict[str, str]] = None):
        """
        Initialize the Excel loader.
        
        Args:
            column_mapping: Optional custom column name mapping
        """
        self.column_mapping = column_mapping or self.DEFAULT_COLUMNS
    
    def load_test_cases(self, file_path: Path) -> List[TestCase]:
        """
        Load test cases from an Excel file.
        
        Args:
            file_path: Path to the Excel file
            
        Returns:
            List of loaded test cases
            
        Raises:
            FileNotFoundError: If file doesn't exist
            ValueError: If file format is invalid
        """
        if not file_path.exists():
            raise FileNotFoundError(f"Test case file not found: {file_path}")
        
        # Load workbook and get first sheet
        workbook = openpyxl.load_workbook(file_path)
        sheet = workbook.active
        
        # Parse rows into test cases
        test_cases = self._parse_sheet(sheet)
        
        workbook.close()
        return test_cases
    
    def _parse_sheet(self, sheet: Worksheet) -> List[TestCase]:
        """
        Parse worksheet into test cases.
        
        Args:
            sheet: The worksheet to parse
            
        Returns:
            List of test cases
        """
        # Get header row and create column index mapping
        headers = [cell.value for cell in sheet[1]]
        col_indices = self._create_column_indices(headers)
        
        # Group rows by Issue Key
        test_case_dict: Dict[str, TestCase] = {}
        
        for row_idx in range(2, sheet.max_row + 1):
            row = sheet[row_idx]
            row_data = self._extract_row_data(row, col_indices)
            
            issue_key = row_data.get("issue_key")
            if not issue_key:
                continue  # Skip rows without Issue Key
            
            # Create or update test case
            if issue_key not in test_case_dict:
                test_case_dict[issue_key] = self._create_test_case(row_data)
            
            # Add test step
            test_step = self._create_test_step(row_data)
            if test_step:
                test_case_dict[issue_key].add_test_step(test_step)
        
        return list(test_case_dict.values())
    
    def _create_column_indices(self, headers: List[Any]) -> Dict[str, int]:
        """
        Create mapping from field names to column indices.
        
        Args:
            headers: List of column headers
            
        Returns:
            Dictionary mapping field names to column indices
        """
        col_indices = {}
        for field_name, column_name in self.column_mapping.items():
            try:
                col_indices[field_name] = headers.index(column_name)
            except ValueError:
                # Column not found, will be None in row data
                pass
        return col_indices
    
    def _extract_row_data(self, row, col_indices: Dict[str, int]) -> Dict[str, str]:
        """
        Extract data from a row based on column indices.
        
        Args:
            row: The worksheet row
            col_indices: Column index mapping
            
        Returns:
            Dictionary of field values
        """
        row_data = {}
        for field_name, col_idx in col_indices.items():
            cell_value = row[col_idx].value
            row_data[field_name] = str(cell_value).strip() if cell_value else None
        return row_data
    
    def _create_test_case(self, row_data: Dict[str, str]) -> TestCase:
        """
        Create a TestCase from row data.
        
        Args:
            row_data: Dictionary of row values
            
        Returns:
            TestCase instance
        """
        return TestCase(
            key=row_data.get("issue_key", ""),
            summary=row_data.get("summary", ""),
            preconditions=row_data.get("preconditions"),
            description=row_data.get("description"),
            status=row_data.get("status"),
            priority=row_data.get("priority"),
            test_case_type=row_data.get("test_case_type"),
            functional_category=row_data.get("functional_category"),
            component_under_test=row_data.get("component"),
        )
    
    def _create_test_step(self, row_data: Dict[str, str]) -> Optional[TestStep]:
        """
        Create a TestStep from row data.
        
        Args:
            row_data: Dictionary of row values
            
        Returns:
            TestStep instance or None if no step data
        """
        step_summary = row_data.get("test_step")
        if not step_summary:
            return None
        
        return TestStep(
            step_summary=step_summary,
            test_data=row_data.get("test_data"),
            expected_result=row_data.get("expected_result"),
        )
